### PR TITLE
Backpack temporary folder

### PR DESCRIPTION
**Pull request by jorgetwgroup**
Dropzone uses this `PurgeTemporaryFolder` job that we `dispatchAfterResponse` whenever a new file is uploaded via ajax. 

Translating, we upload the files, send the response back to the user, we purge older files.

Alternatively user can setup the schedule and the cron job to do that in a general way.

### CODE REVIEW

File: src/app/Http/Controllers/Operations/DropzoneOperation.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
```
Comment (tabacitu) [inline src/app/Http/Controllers/Operations/DropzoneOperation.php:1]: Not sure we should have this trait in CRUD... I understand why it's here, and I know we do this already for some stuff... but we're trying to remove all mentions of PRO inside CRUD...

For the new TrashOperation for example, we didn't do this, we asked people to use `Backpack\PRO\Operations\TrashOperation`...

Yeah no, I don't think we should do this. But let's talk about it, all 3 people, in our meeting later today.

File: webpack.mix.js [hunk 2]
```diff
@@ -75,7 +75,8 @@ mix.copy('node_modules/bootstrap-datepicker/dist', 'src/public/packages/bootstra
 	.copy('node_modules/places.js/dist', 'src/public/packages/places.js/dist')
 	.copy('node_modules/jquery-ui-touch-punch/jquery.ui.touch-punch.min.js', 'src/public/packages/jquery-ui-touch-punch/jquery.ui.touch-punch.min.js')
     .copy('node_modules/urijs/src/URI.min.js', 'src/public/packages/URI.js/URI.min.js')
-    .copy('node_modules/intl-tel-input/build', 'src/public/packages/intl-tel-input/build');
+    .copy('node_modules/intl-tel-input/build', 'src/public/packages/intl-tel-input/build')
+    .copy('node_modules/dropzone/dist', 'src/public/packages/dropzone/dist');
```
Comment (tabacitu) [inline webpack.mix.js:79]: There should also be a `package.json` change here - dropzone becomes a dependency if we do `npm install dropzone` - https://github.com/Laravel-Backpack/CRUD/blob/main/package.json

Comment (jorgetwgroup) [reply webpack.mix.js:79]: Done

**Comment by tabacitu** [comment]
Let's add:
- [ ] config for temp directory
- [ ] config for temp disk
- [ ] command to purge files older than X - `php artisan backpack:purge-temporary-files {--older-than-days}`
- [ ] docs for this feature - it's a Backpack v6 feature

**Comment by pxpm** [comment]
docs here: https://github.com/Laravel-Backpack/docs/pull/426

**Review by tabacitu — CHANGES_REQUESTED** [review]
I am OK with everything that's happening here. Naming, how it's done, everything. Well done 👏👏👏

I only have one request... let's move this to the PRO repo. It feels odd to me that we're introducing this feature in CRUD... when it's not used in CRUD at all. I know we _might_ want to use this temp folder somewhere else, someday... but until then... this is a PRO feature... so it should live there. So I'd move:
- the Command and Job to PRO
- the configs to the `dropzone` operation file

If you think this is odd... that it would limit too much the `temp` directory to the DropzoneOperation... then we could also change the default temp directory from `backpack/temp` to `backpack/temp/dropzone`... that way other addons or features could create their own directory in `backpack/temp`, upload stuff there but they wouldn't get deleted by the purge command. But personally... I don't think that's necessary. I think it's ok to just move them as-is.

What do you guys think? Do you disagree @pxpm or @jorgetwgroup ?

**Comment by pxpm** [comment]
> I am OK with everything that's happening here. Naming, how it's done, everything. Well done 👏👏👏
> 
> I only have one request... let's move this to the PRO repo. It feels odd to me that we're introducing this feature in CRUD... when it's not used in CRUD at all. I know we _might_ want to use this temp folder somewhere else, someday... but until then... this is a PRO feature... so it should live there. So I'd move:
> 
> * the Command and Job to PRO
> * the configs to the `dropzone` operation file
> 
> If you think this is odd... that it would limit too much the `temp` directory to the DropzoneOperation... then we could also change the default temp directory from `backpack/temp` to `backpack/temp/dropzone`... that way other addons or features could create their own directory in `backpack/temp`, upload stuff there but they wouldn't get deleted by the purge command. But personally... I don't think that's necessary. I think it's ok to just move them as-is.
> 
> What do you guys think? Do you disagree @pxpm or @jorgetwgroup ?

Is there a reason to move this to PRO? What is PRO without CRUD ? 

I can give you a reason to don't move it:
- We will endup creating `Backpack\CRUD\PurgeTemporaryFolder` that includes the `Backpack\PRO\PurgeTemporaryFolder` to avoid BC's in future.

Regarding configs I would even go one step further, NO CONFIGS. For what ? A temp folder ? I maybe missing some usecase for temporary folders cause I see no reason for it to be configurable at all.

Temporary folders should only be used as a backup storage in some operation during run-time to avoid polluting the "real storage". You don't expect to work with temporary folders for multiple requests, for that you have the regular storage. 

I setup 72 hours for cleanup in our scenario, but with 2hr of session lifetime (by default in Laravel), anything hanging in that folder past that time (2hrs) is garbage. 

If you still think we should do it, we can add a path or something in the command and in the Job:

`php artisan backpack:purge-temporary-folder --path=dropzone` that we can also be accepted in the job: `new PurgeTemporaryFiles($path = dropzone)` 

You can then configure different cron jobs to clear different subfolders with different times, whatever that means.

**Comment by tabacitu** [comment]
> We will endup creating Backpack\CRUD\PurgeTemporaryFolder that includes the Backpack\PRO\PurgeTemporaryFolder to avoid BC's in future.

Hmm... why would we do that? People can just use `Backpack\PRO\PurgeTemporaryFolder` directly...

> I setup 72 hours for cleanup in our scenario, but with 2hr of session lifetime (by default in Laravel), anything hanging in that folder past that time (2hrs) is garbage.

Fair enough...

> Regarding configs I would even go one step further, NO CONFIGS. For what ? A temp folder ? I maybe missing some usecase for temporary folders cause I see no reason for it to be configurable at all.

Aaah... what about people who don't upload to local... only upload to s3 or something like that? They might not need to customize the `temp_path` but they might need to customize the `temp_disk`. Plus... we want both the DropzoneOperation and the Purge command to point to the same directory... so that directory needs to be stored somewhere... and I think a config is the best place for that, because it also has the benefit of making it configurable.

> If you still think we should do it, we can add a path or something in the command and in the Job: `php artisan backpack:purge-temporary-folder --path=dropzone` that we can also be accepted in the job.

Uuu that's cool. Yes let's do that too! The purge command could accept all the configs - disk, path and time. If none are passed it uses the ones from the dropzone config.

---

I'm still not convinced that it's a good idea to keep it in CRUD. 
Give me more arguments Pedro please, or let's move it to PRO 🙏

**Comment by pxpm** [comment]
> Hmm... why would we do that? People can just use Backpack\PRO\PurgeTemporaryFolder directly...
We add a function to crud (open source) that uses the temporary folder in the future, now pay me to have access to my temporary folder clear command and job. Seems good deal hum? 🙃 

I will do the changes accordingly to your directions 👍 

Cheers

**Pull request closed without merging**